BambooTracker/gui/wave_visual.cpp

```diff
@@ -0,0 +1,75 @@
+#include "gui/wave_visual.hpp"
+#include "gui/color_palette.hpp"
+#include <limits>
+#include <QPainter>
+#include <QDebug>
+
+WaveVisual::WaveVisual(QWidget *parent)
+	: QWidget(parent)
+{
+	setAttribute(Qt::WA_OpaquePaintEvent);
+}
+
+void WaveVisual::setColorPalette(std::weak_ptr<ColorPalette> palette)
+{
+	palette_ = palette;
+}
+
+void WaveVisual::setStereoSamples(const int16_t *buffer, size_t frames)
+{
+	// identify the highest of the 2 input signals
+	int32_t sum = 0;
+	for (size_t i = 0; i < frames; ++i) {
+		sum += std::abs(buffer[2 * i]);
+		sum -= std::abs(buffer[2 * i + 1]);
+	}
+
+	// use this signal as display data
+	samples_.resize(frames);
+	int16_t *samples = samples_.data();
+	for (size_t i = 0; i < frames; ++i)
+		samples[i] = buffer[2 * i + (sum < 0)];
+
+	repaint();
+}
+
+void WaveVisual::paintEvent(QPaintEvent *event)
+{
+	Q_UNUSED(event);
+	QPainter painter(this);
+
+	std::shared_ptr<ColorPalette> palette = palette_.lock();
+	if (!palette)
+		return;
+
+	int w = width();
+	int h = height();
+	painter.fillRect(0, 0, w, h, palette->wavBackColor);
+
+	const int16_t *samples = samples_.data();
+	size_t frames = samples_.size();
+
+	painter.setPen(palette->wavDrawColor);
+
+	int lastY = h / 2;
+	for (int x = 0; x < w; ++x) {
+		size_t index = (size_t)(x * ((double)frames / w));
+
+		int16_t sample = samples[index];
```

Comment:
crash when initial WaveVisual update is happened and samples isn't set to `samples_`. It needs buffer size check before plotting.

Reply:
Added the check `if (frames <= 0)` which prevents entering in this case.